test/test_prototype_transforms_functional.py

```diff
@@ -352,6 +352,29 @@ def vertical_flip_segmentation_mask():
         yield SampleInput(mask)
 
 
+@register_kernel_info_from_sample_inputs_fn
+def pad_segmentation_mask():
+    for mask, padding, padding_mode in itertools.product(
+        make_segmentation_masks(),
+        [[1], [1, 1], [1, 1, 2, 2]],  # padding
+        ["constant", "symmetric", "edge", "reflect"],  # padding mode,
+    ):
+        if padding_mode == "symmetric" and mask.ndim not in [3, 4]:
+            continue
+        if (
+            padding_mode == "edge"
+            and len(padding) == 2
+            and mask.ndim not in [2, 3]
+            or len(padding) == 4
+            and mask.ndim not in [4, 3]
+            or len(padding) == 1
+        ):
+            continue
+        if padding_mode == "edge" and mask.ndim not in [2, 3, 4, 5]:
+            continue
```

Comment:
@federicopozzi33 can you comment out why we skip so much test cases ?

I printed locally which cases it skips and now quite sure about why it is so. For example
```
print(padding_mode, padding, mask.ndim, mask.shape)
>
constant [1] 3 torch.Size([1, 16, 16])
```

Reply:
I fixed. I was skipping too much tests, you're right.

The cases to be skipped are the following.

**First case:**
```
if padding_mode == "symmetric" and mask.ndim not in [2, 3, 4]:
    print(f"skipped: mask_shape={mask.shape}, ndim={mask.ndim}, padding={padding}, padding_mode={padding_mode}")
    continue
```       
Reason: 
`RuntimeError: Symmetric padding of N-D tensors are not supported yet` in `torchvision.transforms.functional_tensor._pad_symmetric`

**Second case:**
```
if (padding_mode == "edge" or padding_mode == "reflect") and mask.ndim not in [2, 3, 4]:
    print(f"skipped: mask_shape={mask.shape}, ndim={mask.ndim}, padding={padding}, padding_mode={padding_mode}")
    continue
```
Reason:
`NotImplementedError: Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now` in `torch.nn.functional.pad`

> So why am I also skipping 5D? 

You have to keep in mind that `len(padding)` is always equal to 4 (since it's parsed before passing it to the torch function), and with that padding the only allowed dimensions are 3 and 4.

> So why am I not skipping 2D?

That because in `torchvision.transforms.functional_tensor.pad`, they are unsqueezed:

```
if img.ndim < 4:
    img = img.unsqueeze(dim=0)
```

Reply:
> mask.ndim not in [2, 3, 4]:

I think `mask` can't have ndim == 2 ? Minimally, it is `(1, H, W)`. cc @pmeier

For N-D cases `e.g. (N, M, K, L, 1, H, W)` we could reshape it into 4D before padding `(N * M * K * L, 1, H, W)` and reshape back inside padding op ?

Reply:
Do you mean something like this?

(The code can be further refactored).
```python
def pad_segmentation_mask(
    segmentation_mask: torch.Tensor, padding: List[int], padding_mode: str = "constant"
) -> torch.Tensor:
  *extra_dims, c, h, w = segmentation_mask.shape 
      
    mask_4d = segmentation_mask.reshape(-1, b, h, w) if len(extra_dims) > 0 else segmentation_mask
    padded = pad_image_tensor(img=mask_4d, padding=padding, fill=0, padding_mode=padding_mode)
      
  *_, new_c, new_h, new_w = padded.shape
  return padded.reshape(*extra_dims, new_c, new_h, new_w) if len(extra_dims) > 0 else padded
```

Seems that `.shape` operation is not supported by `torch.jit`...

```python
cannot statically infer the expected size of a list in this context:
  File "/Users/fpozzi/Documents/OpenSource/my_repos/vision/torchvision/prototype/transforms/functional/_geometry.py", line 402
    segmentation_mask: torch.Tensor, padding: List[int], padding_mode: str = "constant"
) -> torch.Tensor:
    *extra_dims, b, h, w = segmentation_mask.shape 
                           ~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    
    mask_4d = segmentation_mask.reshape(-1, b, h, w) if len(extra_dims) > 0 else segmentation_mask
```

I tried also with .`size()` but I got the same expection.

Reply:
`.shape` is supported by torch script but `*extra_dims` is not.
You can do something like:
```python
mask = torch.rand(5, 4, 1, 32, 32)

def foo(mask):
    c, h, w = mask.shape[-3], mask.shape[-2], mask.shape[-1]
    m = mask.view(-1, c, h, w)
    return m.view(mask.shape)

torch.jit.script(foo)(mask).shape
```

Reply:
> I think `mask` can't have ndim == 2 ? Minimally, it is `(1, H, W)`.

Masks have gotten very little love so far so there is no final decision. But I agree, boolean masks should have shape `(N, H, W)` where `N` is the number of segmentation whether they are individual objects (instance segmentation) or a crowd (semantic segmentation). 

> Do you mean something like this?

Have a look how we are doing it for the bounding boxes:

https://github.com/pytorch/vision/blob/a8f563dbf8520020054aa01f5ae169999775fd19/torchvision/prototype/transforms/functional/_geometry.py#L22-L35

The only difference is that the "data dimensions" are not fixed. For masks this would look like

```py
@torch.jit.script
def pad_segmentation_mask(segmentation_mask: torch.Tensor) -> torch.Tensor:
    shape = segmentation_mask.shape

    num_masks, height, width = shape[-3:]
    segmentation_mask = segmentation_mask.view(-1, num_masks, height, width)
    
    # kernel
    
    return segmentation_mask.view(shape)
```

Reply:
> > Do you mean something like this?
> 
> Have a look how we are doing it for the bounding boxes:
> 
> https://github.com/pytorch/vision/blob/a8f563dbf8520020054aa01f5ae169999775fd19/torchvision/prototype/transforms/functional/_geometry.py#L22-L35
> 
> The only difference is that the "data dimensions" are not fixed. For masks this would look like
> 
> ```python
> @torch.jit.script
> def pad_segmentation_mask(segmentation_mask: torch.Tensor) -> torch.Tensor:
>     shape = segmentation_mask.shape
> 
>     num_masks, height, width = shape[-3:]
>     segmentation_mask = segmentation_mask.view(-1, num_masks, height, width)
>     
>     # kernel
>     
>     return segmentation_mask.view(shape)
> ```

Yeah, the point is that the mask shape before padding is different from mask shape after. 
So I can't simply do:

```python
return segmentation_mask.view(shape)
```

Mask can have any num_dims > 2. So I need to collect dimensions before padding: in particular, I need the extra dimensions, num_masks, height and width.

```python
# Let's suppose that segmentation_mask have shape (3, 4, 1, 5, 5).
extra_dims, num_masks, h, w = segmentation_mask.shape 
print(extra_dims, num_masks, h, w)
> (3, 4), 1, 5, 5
```

Moreover, I haven't figured out yet how to use the `.view` method without the unpacking operator:
```python
# with unpacking operator
padded_mask.view(*extra_dims, num_masks, height, width)
```

Reply:
> Yeah, the point is that the mask shape before padding is different from mask shape after padding.

:facepalm: My bad, sorry. [`resize_image_tensor`](https://github.com/pytorch/vision/blob/e0467c64e337c0d1140a9f9a70a413b7268231f4/torchvision/prototype/transforms/functional/_geometry.py#L38-L54) is a better reference:

```py
@torch.jit.script
def pad_segmentation_mask(segmentation_mask: torch.Tensor) -> torch.Tensor:
    num_masks, old_height, old_width = segmentation_mask.shape[-3:]
    batch_shape = segmentation_mask.shape[:-3]
    segmentation_mask = segmentation_mask.view((-1, num_masks, old_height, old_width))

    # kernel
    new_height, new_width = old_height, old_width

    return segmentation_mask.view(batch_shape + (num_masks, new_height, new_width))
```

Reply:
> > Yeah, the point is that the mask shape before padding is different from mask shape after padding.
> 
> 🤦 My bad, sorry. [`resize_image_tensor`](https://github.com/pytorch/vision/blob/e0467c64e337c0d1140a9f9a70a413b7268231f4/torchvision/prototype/transforms/functional/_geometry.py#L38-L54) is a better reference:
> 
> ```python
> @torch.jit.script
> def pad_segmentation_mask(segmentation_mask: torch.Tensor) -> torch.Tensor:
>     num_masks, old_height, old_width = segmentation_mask.shape[-3:]
>     batch_shape = segmentation_mask.shape[:-3]
>     segmentation_mask = segmentation_mask.view((-1, num_masks, old_height, old_width))
> 
>     # kernel
>     new_height, new_width = old_height, old_width
> 
>     return segmentation_mask.view(batch_shape + (num_masks, new_height, new_width))
> ```

Thank you!

I was missing some things:
First one:

```
size = torch.ones(3, 3).shape[:]
> torch.Size([3, 3])

size_as_tuple = np.ones((3, 3)).shape[:]
> (3, 3)
```

Second one:

```
torch.Size([1, 2]) + (3, 4)
> torch.Size([1, 2, 3, 4])
```